src/nfa/nfa.cc

```diff
@@ -559,7 +601,107 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
+struct StackLevel {
+    State state;
+    Post::const_iterator post_it;
+    Post::const_iterator post_end;
+    StateSet::const_iterator targets_it;
+    StateSet::const_iterator targets_end;
+
+    StackLevel(State q, const Delta & delta):
+            state(q),
+            post_it(delta[q].cbegin()),
+            post_end(delta[q].cend())
+    {
+        if (post_it != post_end) {
+            targets_it = post_it->cbegin();
+            targets_end = post_it->cend();
+        }
+    };
+};
+
+const NumberPredicate<State> Nfa::get_useful_states() const
+{
+#ifdef _STATIC_STRUCTURES_
+    // STATIC SEEMS TO GIVE LIKE 5-10% SPEEDUP
+    static std::vector<StackLevel> stack;
+    //tracking elements seems to cost more than it saves, switching it off
+    static NumberPredicate<State> reached(false);
+    static NumberPredicate<State> reached_and_reaching(false);
+    stack.clear();
+    reached.clear();
+    reached_and_reaching.clear();
+    // STATIC SEEMS TO GIVE LIKE 5-10% SPEEDUP
+    // static std::vector<bool> reached;
+    // static std::vector<bool> reached_and_reaching;
+    // static std::vector<StackLevel> stack;
+    // stack.clear();
+    // reached.assign(size(),false);
+    // reached_and_reaching.assign(size(),false);
+#else
+    std::vector<StackLevel> stack;//the DFS stack
+    //tracking elements seems to cost more than it saves, switching it off
+    NumberPredicate<State> reached(size(),false,false); //what has been reached from the initial states
+    NumberPredicate<State> reached_and_reaching(size(),false,false);//what is reaching final states
+
+    //std::vector<bool> reached(size(),false);
+    //std::vector<bool> reached_and_reaching(size(),false);
+    //std::vector<StackLevel> stack;
+#endif
+
+    for (const State q0: initial) {
+
+        stack.emplace_back(q0,delta);
+        reached.add(q0);
+        //reached[q0]=true;
+        if (final[q0])
+            reached_and_reaching.add(q0);
+            //reached_and_reaching[q0]=true;
+        while (!stack.empty()) {
+            StackLevel & level = stack.back();
+            //continue the iteration through the successors of q (shitty code. A better way, some interface for mata?)
+            while (level.post_it != level.post_end && level.targets_it == level.targets_end) {
+                if (level.targets_it == level.targets_end) {
+                    level.post_it++;
+                    if (level.post_it != level.post_end) {
+                        level.targets_it = level.post_it->cbegin();
+                        level.targets_end = level.post_it->cend();
+                    }
+                } else
+                    level.targets_it++;
+            }
+            if (level.post_it == level.post_end) {
+                stack.pop_back();
+            }
+            else {
+                State succ_state = *(level.targets_it);
+                level.targets_it++;
```

Comment:
`++level.targets_it` could enhance the speed.

Reply:
ok